### ISSUE TITLE
Introduction of Sobolev training capability in Keras

### ISSUE THREAD

**Issue by Nateckert** [keras-team/keras#11493]
Hi,

This is a feature request that I think could be interesting for several users and I wish to share it, as I am not able to implement it.
In the field of scientific computing (especially Computational Fluid Dynamic), there are more and more attempts to use neural network to speed up expensive iterative computations. It can be done in two ways:

- by improving the quality of the initialisation the computation with a neural network

- for computation based on Jacobian evaluation (such as Newton-Raphson), by estimating the Jacobian with a neural network.

In such cases, it is quite common to have access to the derivatives of the function one want to learn, therefore another kind of training can be used : Sobolev training : https://arxiv.org/abs/1706.04859

In such a training, the cost function is not only dependant of the output values of the neural network and the ground truth, but a term depending of the derivatives of the outputs with regards to the inputs is added. This requires to provide also the ground truth derivatives to the learning process.

It would be great to have to possibility to provide the derivatives values in Keras with a call like :

`model.fit(x_train, (y_train, dy_train), epochs=5, batch_size=32, sobolev_coef=0.1)`

`dy_train ` being the matrices of the derivatives of the outputs with regards to the inputs.

The quality of the training is improved, as in addition of the knowledge of the output value, one use also the available knowledge of the derivative of the outputs with regards to the inputs. This would require to compute also for the learning process the derivative of the output of the network with regards to its inputs, in order to be able to do the gradient based update.

Once the learning process is achieved, both outputs values and derivatives values would have been learn in a single neural networks and it accessing to the values of the derivatives would be adding value to the information gathered by such a training, that means being able to call:

`model.predict_derivatives(x_test, batch_size=128)`

Best regards,
